File: MuduoBaseGameServer/src/common/MessageCodec.cc
#include "common/MessageCodec.h"
#include <cstring>
#include <arpa/inet.h>

namespace game {

static int32_t readBigEndianInt32(const uint8_t* buf) {
    return (static_cast<int32_t>(buf[0]) << 24)
         | (static_cast<int32_t>(buf[1]) << 16)
         | (static_cast<int32_t>(buf[2]) << 8)
         |  static_cast<int32_t>(buf[3]);
}

static void writeBigEndianInt32(uint8_t* buf, int32_t value) {
    buf[0] = static_cast<uint8_t>((value >> 24) & 0xFF);
    buf[1] = static_cast<uint8_t>((value >> 16) & 0xFF);
    buf[2] = static_cast<uint8_t>((value >>  8) & 0xFF);
    buf[3] = static_cast<uint8_t>( value        & 0xFF);
}

void MessageCodec::feedBytes(const char* data, size_t len, const MessageCallback& onMessage) {
    // Append to internal buffer
    buffer_.insert(buffer_.end(),
                   reinterpret_cast<const uint8_t*>(data),
                   reinterpret_cast<const uint8_t*>(data) + len);

    // Extract complete frames
    while (buffer_.size() >= 4) {
        int32_t payloadLen = readBigEndianInt32(buffer_.data());
        if (payloadLen < 0 || payloadLen > 10 * 1024 * 1024) { // max 10MB
            // Corrupt frame — reset
            buffer_.clear();
            return;
        }
        size_t frameLen = 4 + static_cast<size_t>(payloadLen);
        if (buffer_.size() < frameLen) {
            break; // Incomplete frame — wait for more data
        }

        // Extract payload
        std::vector<uint8_t> payload(buffer_.begin() + 4, buffer_.begin() + frameLen);

        // Dispatch
        if (onMessage) {
            onMessage(payload);
        }

        // Remove processed frame
        buffer_.erase(buffer_.begin(), buffer_.begin() + frameLen);
    }
}

std::string MessageCodec::encode(const std::string& payload) {
    std::string result;
    int32_t len = static_cast<int32_t>(payload.size());
    uint8_t header[4];
    writeBigEndianInt32(header, len);
    result.append(reinterpret_cast<const char*>(header), 4);
    result.append(payload);
    return result;
}

void MessageCodec::reset() {
    buffer_.clear();
}

} // namespace game

File: unity dev/New Client_Unity_Base_Cli/Assets/Scripts/Network/MessageCodec.cs
using System;
using Game.Messages;
using Google.Protobuf;
using UnityEngine;

namespace Cabo.Client.Network
{
    /// <summary>
    /// Handles [4-byte big-endian length][protobuf bytes] framing.
    /// Accumulates received bytes, extracts complete frames, deserializes ServerMessage.
    /// Also provides static Encode to serialize ClientMessage to bytes for sending.
    /// </summary>
    public sealed class MessageCodec
    {
        private byte[] receiveBuffer = Array.Empty<byte>();

        /// <summary>
        /// Feed received raw bytes into the codec.
        /// Calls onMessage for each complete ServerMessage decoded.
        /// Must be called on the same thread consistently (typically main thread via dispatcher).
        /// </summary>
        public void FeedBytes(byte[] data, Action<ServerMessage> onMessage)
        {
            // Append new data to buffer
            var newBuffer = new byte[receiveBuffer.Length + data.Length];
            if (receiveBuffer.Length > 0)
                Array.Copy(receiveBuffer, 0, newBuffer, 0, receiveBuffer.Length);
            Array.Copy(data, 0, newBuffer, receiveBuffer.Length, data.Length);
            receiveBuffer = newBuffer;

            // Try to extract complete frames
            while (receiveBuffer.Length >= 4)
            {
                int payloadLength = ReadBigEndianInt32(receiveBuffer, 0);
                if (payloadLength < 0)
                {
                    Debug.LogError($"[MessageCodec] Invalid payload length: {payloadLength}, resetting buffer");
                    receiveBuffer = Array.Empty<byte>();
                    return;
                }

                int frameLength = 4 + payloadLength;
                if (receiveBuffer.Length < frameLength)
                    break; // Incomplete frame — wait for more data

                // Extract payload
                var payload = new byte[payloadLength];
                Array.Copy(receiveBuffer, 4, payload, 0, payloadLength);

                // Deserialize
                try
                {
                    var message = ServerMessage.Parser.ParseFrom(payload);
                    onMessage?.Invoke(message);
                }
                catch (Exception ex)
                {
                    Debug.LogError($"[MessageCodec] Parse error: {ex.Message}");
                }

                // Remove processed frame from buffer
                int remaining = receiveBuffer.Length - frameLength;
                if (remaining > 0)
                {
                    var truncated = new byte[remaining];
                    Array.Copy(receiveBuffer, frameLength, truncated, 0, remaining);
                    receiveBuffer = truncated;
                }
                else
                {
                    receiveBuffer = Array.Empty<byte>();
                    break;
                }
            }
        }

        /// <summary>
        /// Serialize a ClientMessage to bytes ready for TCP send.
        /// Returns [4-byte big-endian length][protobuf bytes]
        /// </summary>
        public static byte[] Encode(ClientMessage message)
        {
            var payload = message.ToByteArray();
            var frame = new byte[4 + payload.Length];
            WriteBigEndianInt32(frame, 0, payload.Length);
            Array.Copy(payload, 0, frame, 4, payload.Length);
            return frame;
        }

        /// <summary>
        /// Deserialize raw bytes (full frame payload WITHOUT 4-byte length prefix)
        /// into a ServerMessage.
        /// </summary>
        public static ServerMessage Decode(byte[] payload)
        {
            return ServerMessage.Parser.ParseFrom(payload);
        }

        private static int ReadBigEndianInt32(byte[] buf, int offset)
        {
            return (buf[offset] << 24)
                 | (buf[offset + 1] << 16)
                 | (buf[offset + 2] << 8)
                 | buf[offset + 3];
        }

        private static void WriteBigEndianInt32(byte[] buf, int offset, int value)
        {
            buf[offset]     = (byte)((value >> 24) & 0xFF);
            buf[offset + 1] = (byte)((value >> 16) & 0xFF);
            buf[offset + 2] = (byte)((value >>  8) & 0xFF);
            buf[offset + 3] = (byte)( value        & 0xFF);
        }

        public void Reset()
        {
            receiveBuffer = Array.Empty<byte>();
        }
    }
}
